cmake/bob_robotics.cmake

```diff
@@ -0,0 +1,582 @@
+cmake_minimum_required(VERSION 3.1)
+
+# Build a module from sources in current folder. All *.cc files are compiled
+# into a static library.
+macro(BoB_module)
+    BoB_module_custom(${ARGN})
+    BoB_build()
+endmacro()
+
+# Build a "project" in the current folder (e.g. example, etc.). Each *.cc file
+# found is compiled into a separate executable.
+macro(BoB_project)
+    BoB_init()
+
+    include(CMakeParseArguments)
+    cmake_parse_arguments(PARSED_ARGS
+                          "GENN_CPU_ONLY"
+                          "EXECUTABLE;GENN_MODEL"
+                          "SOURCES;BOB_MODULES;EXTERNAL_LIBS;THIRD_PARTY;PLATFORMS"
+                          "${ARGV}")
+    if(NOT PARSED_ARGS_SOURCES)
+        message(FATAL_ERROR "SOURCES not defined for BoB project")
+    endif()
+
+    # Check we're on a supported platform
+    check_platform(${PARSED_ARGS_PLATFORMS})
+
+    if(PARSED_ARGS_EXECUTABLE)
+        set(NAME ${PARSED_ARGS_EXECUTABLE})
+    else()
+        # Use current folder as project name
+        get_filename_component(NAME "${CMAKE_CURRENT_SOURCE_DIR}" NAME)
+    endif()
+    project(${NAME})
+
+    # Include local *.h files in project. We don't strictly need to do this, but
+    # if we don't then they won't be included in generated Visual Studio
+    # projects.
+    file(GLOB H_FILES "*.h")
+
+    if(PARSED_ARGS_EXECUTABLE)
+        # Build a single executable from these source files
+        add_executable(${NAME} "${PARSED_ARGS_SOURCES}" "${H_FILES}")
+        set(BOB_TARGETS ${NAME})
+    else()
+        # Build each *.cc file as a separate executable
+        foreach(file IN LISTS PARSED_ARGS_SOURCES)
+            get_filename_component(shortname ${file} NAME)
+            string(REGEX REPLACE "\\.[^.]*$" "" target ${shortname})
+            add_executable(${target} "${file}" "${H_FILES}")
+            list(APPEND BOB_TARGETS ${target})
+        endforeach()
+    endif()
+
+    if(PARSED_ARGS_GENN_MODEL)
+        get_filename_component(genn_model_name "${CMAKE_CURRENT_SOURCE_DIR}" NAME)
```

Comment:
Yeah it would make use GeNN 4, as long as the models in ``projects/`` work with it. It sounds like that would also make for simpler CMake code. I opted for basically reimplementing the GeNN makefile stuff in CMake rather than trying to get CMake to play nice with GNU make and I guess that would also make sense for GeNN 4. I suppose it might make sense to write a little CMake package file and have this live in the GeNN repository if it's generic enough. What do you think?

Reply:
yeees that is something of a stumbling block 😄 various variants of the ardin webb one do work, but maybe not the version on master.

I would be all for a tool to simplify building cmake user projects - currently I've written shell/batch(!) scripts to make suitable Makefiles and MSBuild projects e.g. https://github.com/genn-team/genn/blob/genn_4/bin/genn-create-user-project.sh - you could add a -c option to generate cmake instead

Reply:
I'm not super keen on requiring cmake to build the actual generated code though

Reply:
Yeah I agree that getting CMake to call ``genn-buildmodel.sh`` isn't the sensible thing to do for most people (also CMake won't properly trace ``#include``d files because it's not a normal compile), so I wasn't thinking of putting that part in GeNN, though I guess we could have it as a non-default extra option (e.g. by passing ``-DGENERATE_GENN_CODE=1`` to cmake) for people who want to do it in one step. I feel like it does make more sense to auto-generate the code in the context of BoB robotics though.

Reply:
Oh I see what you mean now -- I thought you meant the opposite (i.e. use GeNN to build the generated code, but not actually generate it). We could get CMake to just call ``make`` to build the generated code, which would leave GeNN in charge of setting the right compiler flags etc. (which seems more correct). I guess the other option would be writing a CMake backend for GeNN which doesn't seem hugely appealing

Reply:
That sounds like a good idea - the only issue is that you need to run a different msbuild incantation on Windows

Reply:
(a CMake backend would be a reasonably trivial addition but maybe something for GeNN 4.1)